pkg/minikube/logs/logs.go

```diff
@@ -0,0 +1,141 @@
+/*
+Copyright 2019 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// package logs are convenience methods for fetching logs from a minikube cluster
+package logs
+
+import (
+	"fmt"
+	"os"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/docker/machine/libmachine/log"
+	"github.com/golang/glog"
+	"k8s.io/minikube/pkg/minikube/bootstrapper"
+	"k8s.io/minikube/pkg/minikube/console"
+	"k8s.io/minikube/pkg/minikube/cruntime"
+)
+
+// rootCauseRe is a regular expression that matches known failure root causes
+var rootCauseRe = regexp.MustCompile(`^error: |eviction manager: pods.* evicted|unknown flag: --`)
+
+// importantPods are a list of pods to retrieve logs for, in addition to the bootstrapper logs.
+var importantPods = []string{
+	"k8s_kube-apiserver",
+	"k8s_coredns_coredns",
+	"k8s_kube-scheduler",
+}
+
+// Follow follows logs from multiple files in tail(1) format
+func Follow(r cruntime.Manager, bs bootstrapper.Bootstrapper, runner bootstrapper.CommandRunner) error {
+	cs := []string{}
+	for _, v := range logCommands(r, bs, 0, true) {
+		cs = append(cs, v+" &")
+	}
+	cs = append(cs, "wait")
+	return runner.CombinedOutputTo(strings.Join(cs, " "), os.Stdout)
+}
+
+// IsProblem returns whether this line matches a known problem
+func IsProblem(line string) bool {
+	return rootCauseRe.MatchString(line)
+}
+
+// FindProblems finds possible root causes among the logs
+func FindProblems(r cruntime.Manager, bs bootstrapper.Bootstrapper, runner bootstrapper.CommandRunner) map[string][]string {
+	pMap := map[string][]string{}
+	cmds := logCommands(r, bs, 200, false)
+	for name, cmd := range cmds {
+		log.Infof("Gathering logs for %s ...", name)
+		out, err := runner.CombinedOutput(cmds[name])
+		if err != nil {
+			glog.Warningf("failed %s: %s: %v", name, cmd, err)
+			continue
+		}
+		log.Infof("log length: %d", len(out))
+
+		problems := []string{}
+		for _, l := range strings.Split(out, "\n") {
+			if IsProblem(l) {
```

Comment:
another performance idea: we could do `grep` based on the regex instead on unix systems...that might be faster. 
Alternatively - instead of pulling down all the logs to memory, we could stream it through `Pipe` and process it online...

Reply:
I went with the later approach here. I'd rather keep the processing in Go if possible.